csrc/quantization/int8_kvcache/quant_utils.cuh

```diff
@@ -0,0 +1,284 @@
+// Adated from FasterTransformer, https://github.com/NVIDIA/FasterTransformer/blob/release/v5.3_tag/src/fastertransformer/kernels/decoder_masked_multihead_attention/decoder_masked_multihead_attention_template.hpp
+#pragma once
+
+#include <assert.h>
+#include <stdint.h>
+#include <float.h>
+#include <type_traits>
+#include "../../attention/attention_dtypes.h"
+
+namespace vllm {
+namespace int8 {
+// float32 to int8
+inline __device__ int8_t quant(float a, const float scale, const float zp)
+{
+    int8_t int8;
+    int8 = round(max(-128.f, min(127.f, (a - zp) / scale)));
+    return int8;
+}
+
+// float32x2 to int8x2
+inline __device__ short quant(float2 a, const float scale, const float zp)
+{
+    union {
+        int8_t int8[2];
+        short  int16;
+    };
+
+    int8[0] = quant(a.x, scale, zp);
+    int8[1] = quant(a.y, scale, zp);
+    return int16;
+}
+
+// float32x4 to int8x4
+inline __device__ int32_t quant(float4 a, const float scale, const float zp)
+{
+    union {
+        int8_t  int8[4];
+        int32_t int32;
+    };
+
+    int8[0] = quant(a.x, scale, zp);
+    int8[1] = quant(a.y, scale, zp);
+    int8[2] = quant(a.z, scale, zp);
+    int8[3] = quant(a.w, scale, zp);
+    return int32;
+}
+
+// float16 to int8
+inline __device__ int8_t quant(uint16_t a, const float scale, const float zp)
+{
+    int8_t int8;
+    float  b = half_to_float(a);
```

Comment:
half -> float -> int8. Could data type half be converted directly to int8? Same question on bfloat16.

Reply:
Quant function need `a` to compute with `scale` and `zp` which are`float` type. We convert `a` to `float` to support the computation. So does dequant function.